src/Nonlinear/ReverseAD/Coloring/Coloring.jl

```diff
@@ -0,0 +1,547 @@
+    offsets::Vector{Int}
+    edges::Vector{Tuple{Int,Int}}
+end
+
+_num_vertices(g::UndirectedGraph) = length(g.offsets) - 1
+
+_num_edges(g::UndirectedGraph) = length(g.edges)
+
+_num_neighbors(i::Int, g::UndirectedGraph) = g.offsets[i+1] - g.offsets[i]
+
+_start_neighbors(i::Int, g::UndirectedGraph) = g.offsets[i]
+
+function gen_adjlist(I, J, nel)
+    adjcount = zeros(Int, nel)
+    n_edges = 0
+    for k in 1:length(I)
+        i = I[k]
+        j = J[k]
+        if i == j
+            continue
+        end
+        n_edges += 1
+        adjcount[i] += 1
+        adjcount[j] += 1
+    end
+    offsets = Array{Int}(undef, nel + 1)
+    offsets[1] = 1
+    for k in 1:nel
+        offsets[k+1] = offsets[k] + adjcount[k]
+    end
+    fill!(adjcount, 0)
+    edges = Array{Tuple{Int,Int}}(undef, n_edges)
+    adjlist = Array{Int}(undef, offsets[nel+1] - 1)
+    edgeindex = Array{Int}(undef, length(adjlist))
+    edge_count = 0
+    for k in 1:length(I)
+        i = I[k]
+        j = J[k]
+        if i == j
+            continue
+        end
+        edge_count += 1
+        adjlist[offsets[i]+adjcount[i]] = j
+        edgeindex[offsets[i]+adjcount[i]] = edge_count
+        adjcount[i] += 1
+        adjlist[offsets[j]+adjcount[j]] = i
+        edgeindex[offsets[j]+adjcount[j]] = edge_count
+        adjcount[j] += 1
+        edges[edge_count] = (i, j)
+    end
+    @assert edge_count == n_edges
+    return UndirectedGraph(adjlist, edgeindex, offsets, edges)
+end
+
+struct _Edge
+    index::Int
+    source::Int
+    target::Int
+end
+
+function _prevent_cycle(
+    v,
+    w,
+    x,
+    e_idx1,
+    e_idx2,
+    S,
+    firstVisitToTree,
+    forbiddenColors,
+    color,
+)
+    er = DataStructures.find_root!(S, e_idx2)
+    @inbounds first = firstVisitToTree[er]
+    p = first.source # but this depends on the order?
+    q = first.target
+    @inbounds if p != v
+        firstVisitToTree[er] = _Edge(e_idx1, v, w)
+    elseif q != w
+        forbiddenColors[color[x]] = v
+    end
+    return
+end
+
+function _grow_star(v, w, e_idx, firstNeighbor, color, S)
+    @inbounds e = firstNeighbor[color[w]]
+    p = e.source
+    @inbounds if p != v
+        firstNeighbor[color[w]] = _Edge(e_idx, v, w)
+    else
+        union!(S, e_idx, e.index)
+    end
+    return
+end
+
+function _merge_trees(eg, eg1, S)
+    e1 = DataStructures.find_root!(S, eg)
+    e2 = DataStructures.find_root!(S, eg1)
+    if e1 != e2
+        union!(S, eg, eg1)
+    end
+    return
+end
+
+"""
+    acyclic_coloring(g::UndirectedGraph)
+
+Implement the acyclic coloring algorithm of Gebremdehin, Tarafdar, Manne, and
+Pothen, "New Acyclic and Star Coloring Algorithms with Application to Computing
+Hessians." SIAM J. Sci. Comput. 2007.
+
+Returns `Tuple{Vector{Int},Int}` giving the color index of each node in `g`, as
+well as the total number of colors used.
+"""
+function acyclic_coloring(g::UndirectedGraph)
```

Comment:
I would follow the design from https://github.com/JuliaDiff/SparseDiffTools.jl
and add a ColoringAlgorithm() argument if we want to have several of them implemented (for example copies from SparseDiffTools itself which serve the same purpose)

Reply:
This can all be a second refactoring, after we merge the existing implementation? We should really focus on things that will need breaking changes if we wanted to change them in future.

Reply:
sure ok